jina/peapods/runtimes/request_handlers/data_request_handler.py

```diff
@@ -176,6 +176,9 @@ def handle(
                 # this means the returned DocArray is a completely new one
                 msg.request.docs.clear()
                 msg.request.docs.extend(r_docs)
+        elif partial_requests:
+            msg.request.docs.clear()
```

Comment:
Why do u clear and extend? Shouldn't it be equivalent to keep them?

Reply:
not in the `partial_requests` case.

Reply:
and why do u need to send all the docs. If the meesages that are not the last partial are not forwarded to next Pod?

Reply:
Because it is consistent behavior regarding the Executor. Giving the Executor all Documents inside `docs`, but not forwarding them is confusing.

Reply:
Imagine

```
class MyMergerExecutor):
    @requests
    def merge(self, docs, docs_matrix, **kwargs):
        for partial_docs in docs_matrix:
            for doc in partial_docs:
                assert doc in docs
```

will never fail. Then you expect: Well all Documents will be forwarded, since that is the usual behavior of an Executor.

Reply:
you have to return them as said in the cookbook for now. I think a more agnostic way to do this is to compare r_docs is not Nona and docs != msg.request.docs?

Reply:
Even if it is in the cookbook, I believe it is not a good design. It is rather confusing.

Reply:
ok, if partial_requests is only reached for the last partial message because that is skipped in the `pre_hook`